### PR TITLE
SharedFormInput now supports a placeholder with the label outside variant

### PR DESCRIPTION
**Pull request by rdjpalmer**
## What

Allows the "label outside" variant to have a placeholder without breaking current usage

## Preview

screen shot 2016-03-02 at 10 28 29

### CODE REVIEW

File: components/FormInput/SharedFormInput.js [hunk 1]
```diff
@@ -60,12 +60,13 @@ export default class SharedFormInput extends Component {
...
       label,
+      labelOutside,
       ...remainingProps,
     } = this.props;
     const { focus, id } = this.state;
 
-    const active = value && !!value.length;
-    const isInsideVariant = !!placeholder;
+    const active = labelOutside ? !!placeholder : (value && !!value.length);
+    const isInsideVariant = labelOutside ? false : !!placeholder;
```
Comment (iest) [inline components/FormInput/SharedFormInput.js]: Probably best to rename `active`. This logic isn't easy to follow now we have the extra complication of the `labelOutside` prop.

Comment (rdjpalmer) [inline components/FormInput/SharedFormInput.js]: Suggestion on what to rename it to?

Comment (iest) [inline components/FormInput/SharedFormInput.js]: Not sure, but the logic generally could do with a cleanup so it's more obvious what's happening. 

The `active` className and stuff is only used for the inputs with the labels outside too...

Maybe something more like:
```jsx
const hasPlaceholderAndValue = placeholder && (value && !!value.length);
const labelIsOutside = labelOutside || hasPlaceholderAndValue;

const outerCSS = [
  css.container,
  required
    ? css.required
    : null,
  labelIsOutside
    ? css.labelInside
    : css.labelOutside,
  error
    ? css.error
    : null,
  focus
    ? css.focus
    : null,
  borderless
    ? css.borderless
    : null,
].join(' ');
```

**Comment by rdjpalmer** [comment]
@iest Ready for another review

File: components/FormInput/SharedFormInput.js [hunk 2]
```diff
@@ -60,19 +60,23 @@ export default class SharedFormInput extends Component {
...
+    const labelIsOutsideWithPlaceholder = labelOutside && !!placeholder;
 
     const outerCSS = [
       css.container,
       required ? css.required : null,
-      isInsideVariant ? css.labelInside : css.labelOutside,
+      labelIsOutside ? css.labelOutside : css.labelInside,
+      labelIsOutsideWithPlaceholder ? css.labelOutsideWithPlaceholder : null,
       error ? css.error : null,
-      active ? css.active : null,
```
Comment (iest) [inline components/FormInput/SharedFormInput.js]: This doesn't make sense...

Previously, if the input has a placeholder the input is the inline-label variant.
If not, the label is in place of the placeholder until `active` is true. `active` is only true if a valid value is entered.

So in the CSS: if the input was `active` or was focused on (and is an outside-label variant), the label is moved outside.

With the new behaviour, the label is only moved outside when there's a placeholder and it's an outside-variant.

Inputs that don't have a placeholder now won't work like the used to.

# tl;dr preview

inputs

screenshot 2016-03-03 13 53 33

Comment (iest) [inline components/FormInput/SharedFormInput.js]: A test/spec would be really useful for this 😊

**Comment by rdjpalmer** [comment]
@iest and again

File: components/FormInput/SharedFormInput.js [hunk 3]
```diff
@@ -60,21 +60,46 @@ export default class SharedFormInput extends Component {
...
       css.container,
-      required ? css.required : null,
-      isInsideVariant ? css.labelInside : css.labelOutside,
-      error ? css.error : null,
-      active ? css.active : null,
-      focus ? css.focus : null,
-      borderless ? css.borderless : null,
+      required ?
+        css.required :
+        null,
```
Comment (iest) [inline components/FormInput/SharedFormInput.js]:
```js
required
  ? css.required
  : null,
```
is a nicer pattern IMO (and also happens to be how FB write this pattern from what I've seen, anecdotally).

Comment (rdjpalmer) [inline components/FormInput/SharedFormInput.js]: Hmm, I went with keeping `?` on the same line as it's the indicator of what to expect from the next line, but I'm not fussed. Will change

File: components/FormInput/SharedFormInput.js [hunk 4]
```diff
@@ -60,21 +60,46 @@ export default class SharedFormInput extends Component {
...
+        : null,
+      error
+        ? css.error
+        : null,
+      focus
+        ? css.focus
+        : null,
+      borderless
+        ? css.borderless
+        : null,
```
Comment (iest) [inline components/FormInput/SharedFormInput.js]: Why not:

```js
    const hasValue = value && value.length > 0;
    const isOutsideVariant = labelOutside || !placeholder;
    const labelIsOutside = labelOutside || (hasValue && !placeholder);

    const outerCSS = [
      css.container,
      isOutsideVariant
        ? css.labelOutside
        : css.labelInside,
      labelIsOutside
        ? css.active
        : null,
      required
        ? css.required
        : null,
      error
        ? css.error
        : null,
      focus
        ? css.focus
        : null,
      borderless
        ? css.borderless
        : null,
    ].join(' ');
```


_...assume goodwill_ :dancers:

**Comment by rdjpalmer** [comment]
@iest updated

**Comment by iest** [comment]
:ship:
